src/blas/backends/rocblas/rocblas_helper.hpp

```diff
@@ -0,0 +1,216 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+/**
+ * @file mkl_blas_rocblas.cpp : contains the implementation of all the routines
+ * for rocBLAS backend
+ */
+#ifndef _MKL_BLAS_ROCBLAS_HELPER_HPP_
+#define _MKL_BLAS_ROCBLAS_HELPER_HPP_
+#include <rocblas.h>
+#include <complex>
+#include "oneapi/mkl/types.hpp"
+
+namespace oneapi {
+namespace mkl {
+namespace blas {
+namespace rocblas {
+
+// The static assert to make sure that all index types used in
+// src/oneMKL/backend/rocblas/blas.hpp interface are int64_t
+template <typename... Next>
+struct is_int64 : std::false_type {};
+
+template <typename First>
+struct is_int64<First> : std::is_same<std::int64_t, First> {};
+
+template <typename First, typename... Next>
+struct is_int64<First, Next...>
+        : std::integral_constant<bool, std::is_same<std::int64_t, First>::value &&
+                                           is_int64<Next...>::value> {};
+
+template <typename... T>
+struct Overflow {
+    static void inline check(T...) {}
+};
+
+template <typename Index, typename... T>
+struct Overflow<Index, T...> {
+    static void inline check(Index index, T... next) {
+        if (std::abs(index) >= (1LL << 31)) {
+            throw std::runtime_error(
+                "Rocblas index overflow. rocblas does not support 64 bit integer as "
+                "data size. Thus, the data size should not be greater that maximum "
+                "supported size by 32 bit integer.");
+        }
+        Overflow<T...>::check(next...);
+    }
+};
+
+template <typename Index, typename... Next>
+void overflow_check(Index index, Next... indices) {
+    static_assert(is_int64<Index, Next...>::value, "oneMKL index type must be 64 bit integer.");
+    Overflow<Index, Next...>::check(index, indices...);
+}
+
+class rocblas_error : virtual public std::runtime_error {
+protected:
+    inline const char *rocblas_error_map(rocblas_status error) {
+        switch (error) {
+            case rocblas_status_success: return "rocblas_status_success";
```

Comment:
should we add these in Capital Letters?

Reply:
I am not sure. Let me know what would you prefer.

Reply:
I think we can follow whatever format it is listed in rocBLAS documentation.

Reply:
In the rocBLAS documentation, they snake_case is used: https://rocblas.readthedocs.io/en/latest/API_Reference_Guide.html#_CPPv414rocblas_status

Reply:
Thanks. Then, it makes sense to me to keep it same here.